[1단계 - 출석 구현] 몽이(황재웅) 미션 제출합니다.

wodnd0131 opened the pull request:
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr/studylogs/3925)에 셀프 체크를 작성했나요?


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)
  
### 선택한 점수의 이유를 적어주세요.

시간 관리 부족으로 리펙터링을 통해 보완하지 못한 점이 많기 때문에,
컨밴션을 어느정도 따랐으나 완벽하지 않으며, 객체 지향적인 코드로는 부족하다고 판단하여 2점을 주었습니다.

부족하다고 느낀 점은 다음과 같습니다.

객체 지향을 통해 각 객체 그리고 각 메소드가 독립적으로 작동할 수 있어야 한다고 생각합니다.
그러나 이에 대한 검증이 부족합니다.

또한, 크루 제적 확인을 위한 List와 map이 일급 컬랙션으로 관리되어야 하지만,
이를 리펙터링하지 못한 것처럼 의식했으나 고치지 못한 부분이 적지 않습니다.
ex. handleInput(() -> handleInput()); 와 같이 중복되는 구조가 일부 있음

## 어떤 부분에 집중하여 리뷰해야 할까요?

### 1. 테스트 세분화 및 품질
- 현재 테스트가 단위 테스트로서 충분한지 검토가 필요합니다
- Mocking을 전혀 활용하지 못한 점 보완이 필요합니다

### 2. 도메인 설계 및 캡슐화
- 도메인이 애플리케이션 계층에 노출되지 않도록 하는 전략의 적절성을 검토해야 합니다
- DTO 설계 및 배치의 타당성을 검토해야 합니다:
 - 도메인 패키지 내 DTO 패키지 생성의 적절성
 - 파싱 로직의 도메인 패키지 배치가 올바른지 여부

### 3. 기술적 결정사항 검토
- 서비스 계층의 포맷팅 처리 적절성을 검토해야 합니다
- Map value 업데이트 시 try/catch를 통한 재삽입 로직의 타당성을 검토해야 합니다 (트랜잭션 관리하듯 작성해보았습니다)
- Wrapper 클래스 사용의 적절성을 검토해야 합니다
- Switch-case 사용 제한 이유를 검토해야 합니다
('default'가 'else'와 동일한 맥락이기 때문인가요? 그렇다면 Default가 없다면 가독성을 위해 사용해도 무관하지 않을까요?)

### 4. 아쉬운 싱글톤 패턴 구현
- AttendanceManager의 싱글톤 구현 적절성을 검토해야 합니다
-> 테스트를 위한 초기화 처리 방안이 부재합니다
(Mocking을 위한 리펙터링 시간 관리 부족으로 포기하였습니다)
-> 결국, AttendanceManager의 final 선언을 제거하였습니다

=>final 없이도 AttendanceManager 싱글톤의 불변성은 충족하는지 검토가 필요합니다
(final은 명시적으로 불변성을 보여주는 용도인지 검토가 필요합니다)

### 5. 잘못된 입력 시 재입력 처리
- Supplier를 활용한 재귀적 입력 처리의 안전성이 걱정됩니다
- 해당 파트(View) 관련 재귀로 인한 잠재적 오버플로우 문제의 영향으로 application 자체가 종료되더라도,
Repository와 같이 타 계층에 (적어도 영구적인..) 피해를 주진 않는다고 판단하여 재귀 횟수의 제한을 배제하였습니다.
이는 올바른 선택일지 걱정됩니다.

### 6. 불변성 및 확장성
- 불변성 보장을 통한 예측 가능한 코드 구현 수준이 목표였기에, 얼마나 만족했는지 검토가 필요합니다
- 일급 컬렉션으로의 리팩토링 필요성을 검토해야 합니다 (크루 제적 확인 로직)

Commit:
docs: README.md 작성

Commit:
feat: 출석 후 기록 확인

- 닉네임이 유효하지 않으면 예외가 발생한다.
- 출석 후 출석 기록 확인할 수 있다.

Commit:
fix: 출석을 복수의 날짜를 관리하도록 수정

-Attendance 기존의 출석은 닉네임과 날짜 1:1로 관리했다.

Commit:
feat: 이미 출석한 경우, 예외를 반환

praisebak committed:
feat:  출석 데이터 불러오기 기능 추가

praisebak committed:
test:  테스트용 파일 추가

praisebak committed:
feat: 출석 확인 기능 및 테스트 구현

praisebak committed:
refactor: 컨벤션에 맞지 않은 코드 수정

praisebak committed:
refactor: AttendanceManager의 상수 관리 helper 추가

praisebak committed:
refactor: FileRepository와 AttendanceManager 책임 분리

praisebak committed:
refactor: file resource 패키지 추가

praisebak committed:
feat: 시스템이 2024 12월 데이터들만 사용가능하도록 수정

praisebak committed:
feat: 출석 수정 기능 구현

수정시에 삭제후 삽입하는 방식을 채택함.(트랜잭션과 유사하게)

praisebak committed:
feat: 크루 출석 확인 기능 추가

Commit:
refactor: AttendanceHistoryTest 리펙터링

praisebak committed:
refactor: 컨벤션에 맞지 않던 코드 수정

praisebak committed:
feat: Application 연결 및 버그 수정

praisebak committed:
feat: 제적 위험자 확인 정렬 추가

praisebak committed:
test: 제적 대상자 위험 확인 정렬 테스트 추가

praisebak committed:
docs: 마크다운에 맞춰 수정

jinyoungchoi95 commented on README.md:
```diff
@@ -1,3 +1,44 @@
 # java-attendance
 
-출석 미션 저장소
+# 요구사항   
```
> 1. 테스트 세분화 및 품질
> 현재 테스트가 단위 테스트로서 충분한지 검토 필요
> Mocking을 전혀 활용하지 못한 점 보완 필요

mocking을 왜 활용해야하나요? mocking을 사용하지 않고는 테스트하기 좋은 구조로 만들지 못하는 걸까요? 로또 2단계 미션을 떠올려본다면 mocking을 안쓰고도 테스트할 수 있지 않을까요?

wodnd0131 replied:
Mocking에 대해 고민하는 것은 아직 시기상조였던 부분이네요.
너무 의식적으로 ~~한 경우 ~~ 를 사용해야해! 라 생각한 것 같습니다.
좀 더 현 상황에 맞는 테스트 방식이나 프로덕트 구조에 대해 고민해보겠습니다.🫡

---
### Mocking 고려했던 이유

**AttendanceManager 싱글톤으로 인한 테스트 격리 문제**

- 테스트마다 초기화 필요해 TestAttendanceManager 구현 고려했으나 시간 부족으로 포기
- 대신 final 제거하고 매 테스트마다 초기화로 해결
- _이 과정에서 싱글톤 패턴의 적절성 재고하게 됨_

**Repository 외부 의존성 관련**

- 현재는 Repository에서 읽기만 수행하고 생성자에서 파일 지정 가능
- _실제로는 특별한 Mocking이 불필요했던 것으로 판단_

jinyoungchoi95 commented on src/main/java/attendance/domain/dto/AttendanceDto.java:
> 2. 도메인 설계 및 캡슐화
> 도메인이 애플리케이션 계층에 노출되지 않도록 하는 전략의 적절성
> DTO 설계 및 배치의 타당성:
> 도메인 패키지 내 DTO 패키지 생성의 적절성
> 파싱 로직의 도메인 패키지 배치가 올바른지

배치를 이야기하기전에 dto가 왜 필요한 걸까요? 불변하게 만든 도메인을 그냥 view에 던져주어도 되지 않나요?

wodnd0131 replied:
확실히 Dto가 꼭 필요한가 싶습니다. 
도메인만 최대한 활용할 수 있도록 코드를 수정하였습니다. 

---
초기 DTO 사용 이유와 평가:

1. 매개변수 타입 관리와 가독성 향상이 주 목적
   - 처음엔 파라미터 객체로 구상했으나 DTO 개념으로 발전

2. 팀 내 기준 수립:
   - 비즈니스 로직 없는 바구니 객체는 Record로 구현
   - Application과 Controller 간 데이터 전달에 DTO 사용하여 도메인 은닉

3. 현 프로젝트 관점에서 재평가:
   - 개발 편의성만 고려했고 실질적 이점이 적었음
   - 객체 선언 증가로 리팩토링 부담 증가
   - 민감 정보나 고정된 입출력 양식이 없는 상황에서는 불변성 보장만으로도 충분했을 것
   - 요구사항 대비 과도한 구현이었다고 판단

jinyoungchoi95 commented on src/main/java/attendance/service/CrewDismissService.java:
> 서비스 계층의 포맷팅 처리 적절성

레벨 1 미션은 자바를 어떻게 하면 잘 사용할까를 위한 과정이지 layered architecture을 사용하여 구조화하는 과정이 아닙니다. `Service`, `Controller`, `AppConfig`가 모두 단순히 콘솔 게임 미션에 적절한 역할, 네이밍을 가진 객체가 맞을까요? 알고리즘 문제풀때 controller, service를 만들어가며 푸는 것과 동일해보여요.

스프링에 대한 내용들은 레벨 2에서 배우니 레벨 1에서는 레벨 1에서만 가져갈 수 있는 것을 챙기는데 집중해보세요. 현재는 TDD, 테스트, 클린코드, 객체지향에 대해서 알아갈 수 있는 과정이므로 2, 3, 4 레벨때 뒤늦게 알아가기보다는 레벨 1때 잘 챙겨가면 다음 레벨을 진행하는데도 더 도움이 될 거에요.

> Map value 업데이트 시 try/catch를 통한 재삽입 로직의 타당성 (트랜잭션 관리하듯 작성해봄)

마찬가지로 트랜잭션을 생각할 필요는 없어보입니다.

wodnd0131 replied:
오리의 말에 매우 동의합니다.
때문에, 이번 리펙토링 과정에서 계층 구조를 제거하고, 이 과정을 TDD 사이클에 맞춰서 리펙터링해보았습니다.
이후 커밋을 참고하여 이러한 TDD 학습에 대한 추가적인 피드백 부탁드립니다.

jinyoungchoi95 commented on README.md:
```diff
@@ -1,3 +1,44 @@
+  - 면담 대상자: 결석 3회 이상   
+  - 제적 대상자: 결석 5회 초과   
+
+## 기능 요구사항   
```
> Wrapper 클래스 사용의 적절성

어떤 클래스를 이야기하는지 잘 모르겠는데 어디인지 명확하게 말씀해주시면 같이 확인해보아요.

wodnd0131 replied:
DateTimeFormatterWrapper에 관한 질문이었습니다.

DateTimeFormatter와 같은 라이브러리를 감싸서,
도메인과 무관히 utill 클래스로서 사용하는 것을, 
Wrapper라 지칭해도 괜찮은가? 
그리고 DateTimeFormatterWrapper의 사용은 그 목적에 적절한가? 

이에 관한 질문이었습니다!🫡


~_(이게 도메인패키지에 있는건 잘못된 부분입니다. 🤦‍♂️)_~

jinyoungchoi95 commented on src/main/java/attendance/Application.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    private static void doMethod(AttendanceMethod method) {
+        switch (method) {
```
> Switch-case 사용 제한 이유

추상화를 하는 목적, 객체는 어떻게 생각해야하는가와 엮이어 생각해보면 좋을 것 같아요. 클린코드를 참고해보셔요.

https://lordofkangs.tistory.com/129

wodnd0131 replied:
함수형 메서드를 활용해 선택지를 다루도록 수정했습니다.

---
### Switch-case의 단점?
1. 새로운 타입이 추가될 때마다 switch문을 수정해야 함 -> OCP(Open Closed Principle) 원칙 위반 
2. 여러 책임이 한 곳에 모여있음 -> SRP(Single Responsibility Principle) 원칙

그렇다면, 같은 예시에서 Switch-case 대신 if문을 사용한다고 달라지는가?
-> 당연히 문제점은 동일하다.

---
즉, Switch-case와 if 둘을 비교하기 앞서,
기본적인 설계 타입에 따른 분기 처리 자체가 핵심이라는 것을 알았습니다.

이를 해결하기 위해서,  분기 처리 자체를 막는 것은 요구 사항에서 해결하기 어렵다고 느껴집니다. 때문에 해당 링크에 다룬 다형성 방식을 고민해보았습니다.

그러나, 서비스를 구현하지 않았기에 별도로 클래스에 대한 다형성을 적용하진 않았고,
함수형 메서드를 활용해 분기처리했습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,175 @@
+import java.util.List;
+import java.util.Map;
+
+public class AttendanceManager {
```
> AttendanceManager의 싱글톤 구현 적절성

싱글톤이어야 할 필요가 있는지 모르겠어요. 위에도 말씀드렸듯이 스프링에 대한 이야기는 여기서 하지 않는 것이 맞을 것 같아요.

그걸 제외하고 보면 굳이 `AttendanceManager`가 싱글톤일 필요가 있나요? 둘이상의 `AttendanceManager`를 사용할 수 도 있는 것 아닌가요?

wodnd0131 replied:
확실히 싱글톤은 아직 시기 상조라 생각하여 사용하지 않는 방향으로 수정하였습니다.


---
### AttendanceManager 싱글톤 사용 평가

1. 최초 도입 이유:
   - 계층 구조에서 여러 서비스가 공유하는 객체 의존성 관리
   - 데이터 로딩을 한 번만 수행하고 데이터 일관성 유지 의도

2. 현 프로젝트에서의 문제점:
   - 객체 생성과 데이터 로딩이 분리되어 싱글톤이 데이터 일관성 보장 못함
   - 내부 데이터가 여전히 수정 가능해 불변성 보장 실패
   - 전역 접근성이라는 싱글톤 장점 미활용(매개변수로 전달)
   - 테스트를 위해 final 제거 및 초기화 메서드 추가로 '가짜 싱글톤' 상태

3. 결론:
   - 요구사항에 비해 불필요한 설계였음
   - 데이터 불변성 보장이 목적이었다면 다른 패턴이 더 적합했을 것

jinyoungchoi95 commented on src/main/java/attendance/Application.java:
```diff
@@ -0,0 +1,120 @@
+        });
+    }
+
+    private static <T> T handleInput(Supplier<T> inputSupplier) {
```
> Supplier를 활용한 재귀적 입력 처리의 안전성이 걱정됩니다
> 해당 파트(View) 관련 재귀로 인한 잠재적 오버플로우 문제의 영향 범위가
Repository와 같이 타 계층에 (영구적인) 피해를 주진 않는다고 판단하여 재귀 횟수의 제한을 배제하였습니다

네 현재는 재귀호출에 대해서 그렇게 고민안하셔도 될 것 같아요. 재귀로 버그가 나는 케이스도 없고요. 도메인 객체의 설계가 핵심이니 그쪽을 조금 더 집중해보면 좋을 것 같아요.

jinyoungchoi95 commented on src/test/java/attendance/AttendanceTest.java:
테스트의 네이밍은 `AttendanceTest`인데 실제로 테스트하는 대상이 다르네요.

현재 미션의 가장 큰 목적은 TDD입니다. 객체의 코드를 만들기 전에 테스트를 먼저 작성하여야하는데 커버가 되지 않은 테스트가 많이 보여요.

wodnd0131 replied:
동의합니다. 현재 테스트들이 만족스럽지 않은 점도 있어서,
테스트 케이스를 다시 작성하고 이에 맞게 코드 수정 진행했습니다. 

---
### 테스트 단위 기준 변화:

1. 기존 접근법의 문제:
   - A→B→C→D와 같은 깊은 의존성 구조에서 점진적으로 테스트 범위 확장
   - 하위 기능(D) 테스트 시 상위 기능들(A→B→C)의 정상 작동 전제 필요
   - 결과적으로 단위 테스트가 아닌 통합 테스트 성격으로 변질

2. 개선 방향:
   - A, B, C, D 각 기능을 독립적으로 검증할 수 있는 진정한 단위 테스트로 리팩토링
   - 개별 기능만 검증하여 테스트 복잡성 감소 및 테스트 품질 향상

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceMethod.java:
```diff
@@ -0,0 +1,28 @@
+    private final String method;
+    private static final String INVALID_METHOD = "유효하지 않은 메서드입니다.";
+
+    public static AttendanceMethod of(String requestMethod) {
```
관습적으로 of는 파라메터가 여러개일 때 사용하는 정적 팩토리 메서드명이므로 현재는 from이 맞습니다.

왜 from과 of의 사용케이스 차이가 나는걸까요?

wodnd0131 replied:
>여러 개별 값들로부터 생성한다면 -> of 사용
>다른 타입의 객체로부터 변환한다면 -> from 사용

따라서 인스턴스를 생성할 때, 위과 같이 사용 케이스로 나뉘어집니다.
이에 맞춰 코드 수정하도록 하겠습니다.🫡

---
### from과 of의 사용케이스 차이가 나는 이유는 다음과 같습니다.

1. of는 여러 매개변수를 받아서 인스턴스를 생성할 때 사용합니다.
   - of는 여러 독립적인 값들을 나열할 때 자연스럽게 읽힙니다. ( 예를 들면, List of elements )

2. from은 단일 매개변수를 통해 타입 변환이 필요한 경우에 주로 사용합니다.
   -  from은 타입 변환의 의미를 더 명확하게 전달합니다. "Create date from instant"와 같이 변환의 출발점을 명확히 할 수 있는 영어 문법적인 부분이 작용합니다.

3. Java 8부터 이러한 컨벤션이 표준 라이브러리에 일관되게 적용되어 왔습니다. 
개발자들이 API를 예측 가능하게 사용할 수 있는 일관성을 가지고 있습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceDismiss.java:
```diff
@@ -0,0 +1,27 @@
+package attendance.domain;
+
+public class AttendanceDismiss {
```
static 메소드만 있는데 `AttendanceDismiss`를 도메인 객체라고 부를 수 있을까요?

몽이가 정의하는 "도메인"은 무엇인가요?

wodnd0131 replied:
도메인은 '비즈니스 로직을 가진 객체'입니다. 
여기서 객체란 상태와 행동을 모두 가지는 것이 핵심 특징입니다.
AttendanceDismiss는 상태가 없고 단순 작업의 집합이기에 도메인으로 보기 어렵습니다. 행동을 통해 자신의 상태를 관리하는 객체지향의 캡슐화와 달리, 정적 메소드 집합에 불과합니다.

이러한 경우 유틸리티 클래스나 서비스로 분류하는 것이 더 적합하겠네요.🫡

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,175 @@
+        instance = null;
+    }
+
+    private HashMap<String, Attendances> attendanceManager = new HashMap<>();
+    private String ATTENDANCE_RESULT_FORMAT = "%s (%s)";
```
```suggestion
    private static final String ATTENDANCE_RESULT_FORMAT = "%s (%s)";
    private final HashMap<String, Attendances> attendanceManager = new HashMap<>();
```

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceManagerHelper.java:
```diff
@@ -0,0 +1,24 @@
+import java.time.LocalDate;
+import java.time.LocalTime;
+
+public class AttendanceManagerHelper {
```
`Helper`가 도메인이 맞나요? constant를 모아두는 클래스를 만들기보다는 각 상수들을 적절한 객체에서 가지고 활용할 수 있도록 변경해보세요.

wodnd0131 replied:
AttendanceManager에 너무 많은 상수가 포함되어 분리했던 부분입니다.
사실 이는 다시 말해 너무 많은 책임이 있는 거기도 하죠.
AttendanceManager를 분리하고, 각각 객체 내부에서 enum으로 관리하겠습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceMethod.java:
```diff
@@ -0,0 +1,28 @@
+import attendance.exception.AttendanceArgumentException;
+import java.util.Arrays;
+
+public enum AttendanceMethod {
```
화면의 요구사항이 1, 2, 3, 4, Q가 아닌 다른 명령으로 변경되면 화면의 요구사항으로 인해 도메인 코드가 변경되는 영향이 있겠네요.

wodnd0131 replied:
Application 단에서만 수정이 일어나도록 제한하여, 기존 도메인에 대해서는 변경되지 않도록 막았습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.Map;
+
+public class Attendances {
+    private final Map<LocalDate, Attendance> attendances = new HashMap<>();
+    private static final String DUPLICATE_ATTENDANCE_DATE = "이미 출석되었습니다. 수정 기능을 이용해주세요.";
+    private static final String NOT_EXIST_ATTENDANCE = "해당 날짜에 출석이 존재하지 않습니다.";
```
```suggestion
    private static final String DUPLICATE_ATTENDANCE_DATE = "이미 출석되었습니다. 수정 기능을 이용해주세요.";
    private static final String NOT_EXIST_ATTENDANCE = "해당 날짜에 출석이 존재하지 않습니다.";
    private final Map<LocalDate, Attendance> attendances = new HashMap<>();
```

자바 컨벤션상 상수가 위에 위치하여야 합니다.

https://www.oracle.com/java/technologies/javase/codeconventions-fileorganization.html

jinyoungchoi95 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private LocalTime determineAttendanceStartTime(LocalDate currentDate) {
+        if (currentDate.getDayOfWeek().getValue() == AttendanceManagerHelper.MONDAY) {
```
```suggestion
        if (currentDate.getDayOfWeek() == DayOfWeek.MONDAY) {
```
여도 충분하지 않을까요?

jinyoungchoi95 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,67 @@
+        attendances.put(currentDate, attendance);
+    }
+
+    public LocalTime getAttendanceTime(LocalDate datetime) {
+        return attendances.get(datetime)
+                .time();
+    }
+
+    public AttendanceStatus getAttendanceStatus(LocalDate date) {
+        return attendances.get(date)
+                .attendanceStatus();
+    }
```
날짜에 존재하는 출석값이 없으면 어떻게 되나요?

wodnd0131 replied:
제적 대상자 조회 등, 전체 조회할 경우, 결석에 대한 통계를 얻고자 할 때 해당 작업에 한하여 추가합니다.
예를 들어 날짜가 없을 경우, 이를 체크하고 return 값으로 날짜 없을 때 필요한 값을 전달합니다.
( --:-- (결석)와 같은 )
-> Attendance를 통해 따로 저장 및 관리 않음

jinyoungchoi95 commented on src/main/java/attendance/exception/AttendanceStateException.java:
```diff
@@ -0,0 +1,8 @@
+package attendance.exception;
+
+public class AttendanceStateException extends IllegalStateException {
```
사용하지 않는 클래스는 모두 지워주세요.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 몽이 😄

출석 미션의 리뷰어를 맡은 오리입니다.

첫 미션 잘 구현해주셨네요. 다음 단계 넘어가기 전 같이 이야기하면 좋을 내용 코멘트 남겨두었으니 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

Commit:
refactor: 미사용 코드 제거 및 컨밴션 적용

Commit:
refactor: 단위 테스트를 위한 테스트 케이스 재작성

-단위 테스트를 만족하지 못했던 기존의 Test 코드들을 삭제한다.
-기능에 맞는 테스트 케이스를 작성한다.
-불필요해진 테스트용 파일들을 제거한다.

Commit:
docs: README.md에 리펙터링 목표 추가

-피드백을 위한 리펙터링 순서 및 목표를 작성한다.

Commit:
refactor: 불필요한 계층 구조 제거

Commit:
feat: 등교 시간을 입력하면, 출석한다

- Attendance를 통해 등교 시간과 출석 상태를 관리한다.
- 닉네임에 관한 책임은 AttendanceManger가 가진다.

Commit:
feat: 출석 시간보다 5분 초과되어 출석할 때, 지각 처리한다

Commit:
feat: 출석 시간보다 30분 초과되어 출석할 때, 결석 처리한다

Commit:
refactor: 출석 상태 결정 로직 가독성 개선

Commit:
feat: 월요일은 별도의 규칙으로 출석 상태를 결정한다

- test:월요일에 출석할 때, 교육시간이 13시부터이다.
- test:월요일은 13:05 이후 출석할 경우, 지각 처리된다.
- test:월요일은 13:30 이후 출석할 경우, 결석 처리된다.

Commit:
feat: 출석하는 날짜가 주말인 경우, 예외가 발생한다

Commit:
feat: 캠퍼스 운영시간 외에 출석할 경우, 예외가 발생한다

Commit:
refactor: Attendance validate 및 상수 관리 개선

Commit:
test: 출석하는 날짜가 공휴일인 경우, 예외가 발생한다

-@Disable 처리된 테스트 수정

Commit:
refactor: Attendance의 indent 개선

Commit:
feat: AttendanceManager에서 파일 불러오기 및 예외 처리

-잘못된 File에 대한 예외 케이스 테스트를 추가한다
-불러온 데이터에 대한 처리는 포함하지 않는다

Commit:
feat: AttendanceManager에서 파일 정보 및 출석 정보 저장

-test: csv 파일로부터 출석 정보를 불러온다
-test: 닉네임과 출석 정보을 입력하면, 출석 정보를 저장한다

Commit:
refactor: AttendanceManager의 상수를 내부 클래스로 관리

Commit:
feat: 등록되지 않은 닉네임을 입력하면, 예외가 발생된다

Commit:
feat: 다시 출석할 경우, 예외가 발생한다

Commit:
refactor: List<Attendance>를 일급컬랙션 객체 AttendanceList로 포장

Commit:
refactor: AttendanceManager의 file에 대한 책임 분리

Commit:
refactor: AttendanceManager를 추상 클래스로 변경

-Map에 대한 일급 컬랙션 객체, AttendanceBook 생성
-AttendanceList를 record 클래스로 변경
-Attendance를 상속받은 AttendanceRegister 생성
-마찬가지로, AttendanceModifier 생성

Commit:
refactor: AttendanceManager.manage 파라미터 개선 및 패키지 정리

Commit:
feat: 닉네임과 날짜, 수정 시간을 입력한 후, 출석을 수정한다

-AttendanceList의 list를 static하지 않도록 수정

Commit:
feat: 등록되지 않은 닉네임을 입력할 경우, 예외가 발생한다

Commit:
feat: 출석하지 않은 날에 대한 출석을 수정한다

Commit:
test: 출석이 수정될 경우, 출석 상태도 수정한다

-test: 출석 상태를, 지각으로 수정한다
-test: 출석 상태를, 결석으로 수정한다
-test: 월요일일 경우에 맞게 출석을 수정한다

Commit:
test: 출석 날짜 및 시간이 올바르지 않을 경우, 예외를 발생한다

-test: 수정 시간이 캠퍼스 운영 시간 이외일 경우, 예외를 발생한다
-test: 수정 날짜가 주말일 경우, 예외를 발생한다
-test: 수정하는 날짜가 공휴일인 경우, 예외를 발생한다

Commit:
feat: 닉네임을 입력하면, 크루 출석 기록을 확인한다

Commit:
test: 닉네임을 입력하면, 크루 출석 기록을 확인한다

-test: 등교하지 않은 날에 대해서도 출석 기록에 포함한다
-test: 등록되지 않은 닉네임을 입력할 경우, 예외가 발생한다

Commit:
test: 해당 크루의 출석 상태 통계를 확인할 수 있다

Commit:
refactor: 가독성 개선 및 불필요한 코드 제거

Commit:
refactor: 출석 상태 통계를 정렬하여 반환

Commit:
feat: 해당 크루의 제재 수준을 판단한다

Commit:
feat: 모든 크루원의 제재 통계를 출력한다

Commit:
refactor: 불필요한 코드 제거 및 AttendanceManager.getResult() 추상화

Commit:
test: 제재 상태 정렬 테스트

- csv 변경에 따른 테스트 input 조정

Commit:
refactor: 클래스명 조정

-refactor: 도메인의 일반 클래스에서 생성자 대신 정적 생성자 사용
-refactor: 정적 클래스에 대한 private 생성자 적용

Commit:
refactor: AttendanceManager 관련 추상화 제거

Commit:
refactor: RegisterManager 출력 관련 리펙터링

Commit:
refactor: ModifyManager 출력 관련 리펙터링

Commit:
refactor: Application 리펙터링

-refactor: 입력값 형식 검증 추가

Commit:
refactor: StatusStatistic 코드 개선

-refactor: enum 계산식 수정
-refactor: compareTo 개선

Commit:
refactor: 불필요한 코드 제거 및 컨벤션 적용

Commit:
feat: 기능별 출력 메세지 추가

Commit:
refactor: OutputView 상수 개선

Commit:
docs: README.md 리펙터링 개요 추가

jinyoungchoi95 commented on src/main/java/attendance/common/SystemDateConfig.java:
Config라는 것 또한 미션에서 필요없고, 상수는 모두 적정 도메인에서 관리하도록 두세요. 상수를 관리하는 클래스는 객체지향일까요?

wodnd0131 replied:
객체 지향에서는 데이터가 아닌 객체를 취급해야 합니다.
그래서 분리된 상수 또한 객체로 볼 수 있음을 깨달았습니다.

지금까지 상수를 분리한 것은, 단순히 둘 이상의 객체에서 공통으로 사용했기 때문입니다.
이는 둘 이상의 곳에서 공통된 도메인 로직이 들어감을 뜻합니다.

다시 말해, 상수에 대한 책임이 있는 로직이 있다는 것입니다.
이는 기존의 도메인 객체들의 책임으로부터 분리할 수 있음을 느꼈습니다.
그래서, 분리된 상수들을 관리할 수 있는 객체를 만들고, 기존의 객체들이 새로 만든 객체들에게 상수에 관한 업무를 요청하도록 수정하였습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/attendance/Attendance.java:
```diff
@@ -0,0 +1,108 @@
+        }
+    }
+
+    private static void validateIsDuringCampusSchedule(LocalDateTime dateTime) {
+        LocalTime time = dateTime.toLocalTime();
+        if (isDuringCampusSchedule(time)) {
+            throw new AttendanceArgumentException(Message.OUT_OF_SCHOOL_SCHEDULE);
+        }
+    }
+
+    private static boolean isDuringCampusSchedule(LocalTime time) {
+        return time.isBefore(Schedule.CAMPUS_OPEN) || time.isAfter(Schedule.CAMPUS_CLOSE);
+    }
```
```suggestion
    private static void validateIsDuringCampusSchedule(LocalDateTime dateTime) {
        LocalTime time = dateTime.toLocalTime();
        if (time.isBefore(Schedule.CAMPUS_OPEN) || time.isAfter(Schedule.CAMPUS_CLOSE)) {
            throw new AttendanceArgumentException(Message.OUT_OF_SCHOOL_SCHEDULE);
        }
    }
```
if문 하나만 바라보아도 되는 상황에서 굳이 메소드로 분리할 필요는 없어보여요.

jinyoungchoi95 commented on src/main/java/attendance/domain/attendance/Attendance.java:
```diff
@@ -0,0 +1,108 @@
+        return dateTime.toLocalDate().equals(date);
+    }
+
+    private static class Schedule {
```
상수를 관리하는 클래스는 모두 지우고, 도메인들이 관리하도록 해보세요.

jinyoungchoi95 commented on src/main/java/attendance/domain/attendance/AttendanceBook.java:
```diff
@@ -0,0 +1,74 @@
+
+public record AttendanceBook(Map<String, AttendanceList> attendances, List<StatusStatistic> statusStatistics) {
+
+    public static AttendanceBook from(List<String> lines) {
```
현재 정적 팩토리 메서드를 사용함으로써 얻는 장점이 있나요?

wodnd0131 replied:
기존에는 크게 의미를 두지 않고, 단순하게 정적 팩토리 메소드를 활용했던 것 같습니다.

이번 리펙토리에서는 기본 생성자와 다른 파라미터존에는 크게 의미를 두지 않고, 단순하게 정적 팩토리 메소드를 활용했던 것 같습니다.


이번 리펙토리에서는 다음과 같은 기준으로 사용했습니다.
기본 생성자와 다른 파라미터를 가질 경우,  생성에 대한 로직을 포함하게 됩니다.
생성에 대한 책임이 생기므로, 이 경우 정적 팩토리 생성자를 사용했습니다.

일반 생성자에 구체적인 로직을 담는 것보다 상황에 따라 네이밍 등도 조절할 수 있는 정적 팩토리 메서드가 
유지보수에 좀 더 유리하다고  생각했기 때문입니다.

또한, AttendanceBook 같은 경우, 기존의 생성 시 Attendances와 Attendance를 모두 직접 선언하였습니다.
이는 너무 강한 의존을 만드는 것 같아서, 별도의 팩토리 클래스로 분리하였습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/attendance/AttendanceBook.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    public AttendanceList getAttendanceList(String nickname) {
+        var attendanceList = attendances.get(nickname);
+        if (attendanceList == null) {
+            throw new AttendanceArgumentException(Constant.NOT_REGISTERED_NICKNAME);
+        }
+        return attendanceList;
```
```suggestion
        if (!attendances.containsKey(nickname)) {
            throw new AttendanceArgumentException(Constant.NOT_REGISTERED_NICKNAME);
        }
        return attendances.get(nickname);
```
로 표현될 수 있겠네요.

jinyoungchoi95 commented on src/test/java/attendance/AttendanceBookTest.java:
```diff
@@ -0,0 +1,53 @@
+import attendance.domain.attendance.AttendanceBook;
+
+public class AttendanceBookTest {
+    private AttendanceBook manager;
+    private static final String TEST_FILE = "/attendances.csv";
+
+    @BeforeEach
+    void setUp() throws AttendanceFileException {
+        var repository = AttendanceFileReader.from(TEST_FILE);
+        var lines = repository.getLines();
+        manager = AttendanceBook.from(lines);
+    }
```
```suggestion
    private final AttendanceFileReader reader = AttendanceFileReader.from("/attendances.csv"); 
    private final AttendanceBook manager = AttendanceBook.from(reader.getLines());
```
1. junit5의 기본 라이프사이클은 메소드이기 때문에 beforeEach로 초기화할 필요는 없습니다.
https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-instance-lifecycle

2. repository가 아닌 reader아닌가요?

wodnd0131 replied:
이름을 수정하는 리펙터링 과정에서 실수가 있었네요..

repository라는 이름에 맞는 기능을 하는가? 다시 고민해보았습니다.

파일에 대한 CRUD에서도 단순히 R만 가졌으며, 도메인을 직접 생성하지 않았기에,
단순히 csv를 읽어오는 역할로 CsvReader로 네이밍을 수정했습니다.

도메인에 관한 부분이 없다고 생각하여, Utilliy에 배치하였는데 오리는 어떻게 생각하시나요?
오히려 도메인 생성의 책임을 주고, 도메인 객체로 관리하는 것이 더 좋았을까요?🤔

jinyoungchoi95 replied:
네 말씀해주신 이유로 util로도 현재는 충분하다고 생각해요.

jinyoungchoi95 commented on src/main/java/attendance/domain/attendanceManager/ModifyManager.java:
```diff
@@ -0,0 +1,70 @@
+    private final AttendanceBook attendanceBook;
+    private final StringBuilder report;
+
+    public ModifyManager(AttendanceBook attendanceBook) {
+        this.attendanceBook = attendanceBook;
+        this.report = new StringBuilder();
+    }
+
+    public void manage(String nickname, LocalDateTime dateTime) {
+        Attendance newAttendance = new Attendance(dateTime);
+        AttendanceList attendanceList = attendanceBook.getAttendanceList(nickname);
+        Optional<Attendance> existingAttendanceOpt = attendanceBook.findAttendance(nickname, dateTime.toLocalDate());
```
> 줄여 쓰지 않는다(축약 금지).

Opt가 무엇인가요?

jinyoungchoi95 commented on src/main/java/attendance/domain/attendanceManager/ModifyManager.java:
```diff
@@ -0,0 +1,70 @@
+        appendNewAttendanceInfo(newAttendance);
+    }
+
+    private void appendExistingAttendanceInfo(Attendance attendance) {
```
> 클린코드 Chapter2. 의미 있는 이름

클래스 이름에 Info, Data와 같은 불용어를 사용하지 않는다.

jinyoungchoi95 commented on src/main/java/attendance/domain/attendanceManager/ModifyManager.java:
```diff
@@ -0,0 +1,70 @@
+
+    public void manage(String nickname, LocalDateTime dateTime) {
+        Attendance newAttendance = new Attendance(dateTime);
+        AttendanceList attendanceList = attendanceBook.getAttendanceList(nickname);
```
> 클린코드 Chapter2. 의미 있는 이름
List와 같은 개발자에게만 의미를 갖는 이름을 변수명에 사용하지 않습니다.

추가로 복수를 의미하는데 자료구조의 구현내용을 담으면 자료구조가 Set이 되면 네이밍이 수정되어야하는 불편함을 겪어야겠네요.

jinyoungchoi95 commented on src/main/java/attendance/domain/attendance/AttendanceList.java:
몽이가 생각하는 일급컬랙션은 무엇인가요? 단순히 `List`들을 감싼 객체인가요? 

List라는 네이밍이 들어가야하는 이유가 무엇인가요?

wodnd0131 replied:
일급 컬랙션은 컬랙션을 관리하기 위해서 사용합니다.

이때 관리한다는 다음과 같은 의미입니다.
- 연관된 상태와 행위를 한 곳에서 관리합니다. 즉, 객체로서 관리합니다.
- 도메인에 특화된 컬렉션을 만들 수 있습니다.
- 컬렉션에 이름을 부여함으로써 의도를 명확히 표현할 수 있습니다.
- 컬렉션의 불변성을 보장할 수 있습니다.

이 중 무엇보다, 객체로 관리해야 한다는 의미를 좀 더 상기하게 되었습니다.
객체이기에 그 역할을 알 수 있는 의미있는 이름있어야 하고,
그 역할에 책임을 다 할 수 있는 도메인에 특화된 기능을 가지는 것이니까요.

그런 면에서 전반적으로 네이밍이 부족한 면이 객체에 대한 고민이 적어서 그런 것 같습니다.
네이밍에 관해서도 객체를 잘 나타내는 것이 가지는 의미를 고민하겠습니다.🫡

jinyoungchoi95 commented on src/main/java/attendance/domain/attendance/AttendanceList.java:
```diff
@@ -0,0 +1,129 @@
+    private final List<Attendance> attendances = new ArrayList<>();
+    private final List<String> attendanceHistory = new ArrayList<>();
+
+    public void add(Attendance attendance) {
+        attendances.add(attendance);
```
똑같은 시간의 출석정보가 중복해서 들어오는 것을 허용하나요?

wodnd0131 replied:
로직을 통해 막아 두었지만, 
이는 중복에 대한 책임을 다른 객체에 일임했다고 다시 생각하게 되네요. 
해당 객체가 스스로 중복을 관리하도록 수정 하겠습니다.

최종적으로 List -> Map으로 바꾸었습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/attendance/AttendanceList.java:
```diff
@@ -0,0 +1,129 @@
+
+    public Attendance findAttendance(Attendance attendance) {
+        return attendances.stream()
+            .filter((i) -> i.equals(attendance))
```
> 줄여 쓰지 않는다(축약 금지).
i로 축약하지 않습니다. 나머지 코드도 모두 동일합니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/attendance/AttendanceList.java:
```diff
@@ -0,0 +1,129 @@
+    public Map<AttendanceStatus, Integer> produceStatistic() {
+        Map<AttendanceStatus, Integer> statistics = new HashMap<>();
+        List<LocalDate> workingDays = extractWorkingDays();
+        workingDays.forEach(date -> processAttendanceForDate(date, statistics));
```
stream의 foreach 종단 연산은 안티패턴입니다.

https://tecoble.techcourse.co.kr/post/2020-05-14-foreach-vs-forloop/

jinyoungchoi95 commented on src/main/java/attendance/domain/attendanceManager/SanctionManager.java:
```diff
@@ -0,0 +1,34 @@
+import attendance.domain.attendance.AttendanceBook;
+
+public class SanctionManager {
+    public static final String INIT_MESSAGE = "\n제적 위험자 조회 결과\n";
```
화면에 출력되는 모든 것은 view의 요구사항입니다. domain의 정의가 무엇인가요?

wodnd0131 replied:
도메인은 해결하고자 하는 문제 영역입니다.
비즈니스 로직과 규칙을 객체로 표현하는 것이며,
View와 같이 도메인이 아닌 로직을 나누어 관심사를 분리합니다.

이때, 도메인 로직을 나누는 기준은
"이 로직이 요구사항에 직접 관련되어있는가?"
"이 로직이 변경되면 비즈니스 규칙이 변경되는가?"
등으로 비도메인 로직과 구분할 수 있습니다.

그렇기에, 출력에 관한 부분이나 DB와 같은 기술 부분은 비 도메인 로직으로 보며, 이를 분리해야 합니다.

이런 도메인 로직에 대한 기준이 확고하지 않아 다음과 같이 고민했었습니다.

"화면에 출력되는 것은 View 책임이지만, View에게 전달되는 데이터는 도메인에서 만들어야 하지 않을까?
그렇다면, View에서 출력하는 과정이 도메인과 분리되어있어야 하니, 바로 출력할 수 있는 값을 보내야하나?"

하지만, 이러한 로직들은 앞서 기준으로 다루었듯이, 해당 도메인에게 요구되는 책임은 아닙니다.
그래서, 결국 누가하냐의 고민이였던 것 같습니다.

현재 코드에서는 어플리케이션에서 진행하도록 수정했지만, 개선할 필요성이 있는 것 같습니다.

다음 리펙토링에서, EnumTextConverter와 같이 인터페이스로 출력 형태를 정의하는 것도 시도해 보고자 합니다.
이는 괜찮은 접근일까요?

jinyoungchoi95 replied:
DB는 현재 생각할 관심사가 아닌 것 같아요. 현재 배우는 단계에서는 화면과 도메인 그자체에서만 신경써서 어떻게 구분지을지 고민해보시는게 학습 포인트입니다.

출력의 형태에 대해서는 너무 고민하지마시고 도메인을 어떻게 잘 설계할지부터 고민해보세요. 대부분의 pr을 보면 과하게 다른것을 고민하다 도메인 설계를 놓치는데 도메인을 정말 리뷰할것없이 완벽하게 구성해보는 걸 도전해보세요. 도메인에 객체지향, TDD, 클린코드 등 현재는 그것만으로도 공부해야하는 양은 충분히 많은 것 같습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,17 @@
+package attendance.domain;
+
+public enum AttendanceStatus {
+    ATTENDANCE("출석"),
+    LATE("지각"),
+    ABSENCE("결석");
```
여기저기 도메인과 화면의 요구사항이 섞여있네요. 레벨1 미션에서는 domain 패키지 내에 있는 모든 화면과 관련된 기능들은 제거하세요.

wodnd0131 replied:
도메인 쪽 enum에 String값을 지우고,
View 쪽에 같은 배열에 String 값을 가진 enum을 만들어 
도메인의 Enum 전달받으면, Enum에 따라 String 값을 반환하도록 하였습니다.

이를 통해 Enum 클래스들에 대해 도메인의 로직에 의한 변화와 
출력 형태(예를들면 영문출력)에 대한 변화를 분리하였습니다.  

서비스에서 static으로 변환을 호출하기 때문에
수정해야 하는 곳이 더 늘었다는 단점이 생겨서 좀 더 개선할 필요성이 있었습니다.

그래서 리펙터링된 코드처럼 enum에서 인터페이스를 통해 도메인에서의 사용 형태를 고정하고,
String 변환에 대한 작업 변경은 EnumTextConverter에서만 수정하도록 하였습니다.

jinyoungchoi95 commented on src/main/java/attendance/common/exception/AttendanceArgumentException.java:
common이 무엇인가요? 현재 미션은 출석에 대한 기능밖에 없는데 공통화해서 다른 곳에서 사용중인 프로그램이 있나요?

wodnd0131 replied:
앞서 분리한 공통 상수들의 문제처럼, Common의 내용물들이 필요하지 않은 상황이었네요.
습관처럼 사용해와서 고민 없이 편한 선택을 해버린 것 같습니다.🤦‍♂️

jinyoungchoi95 commented on src/main/java/attendance/domain/StatusStatistic.java:
```diff
@@ -0,0 +1,105 @@
+            .compare(this, o);
+    }
+
+    public enum SanctionLevel {
```
클래스 내부에 선언되어야하는 이유가 있나요?

wodnd0131 replied:
공통 상수를 어떻게 객체로 관리할 수 있지? 라는 생각에서 지나온 과도기이지만,,
내부에 숨겼을 뿐 이전과 큰 차이도 없습니다.🤔

상수에 대하여 가장 적합한 객체를 찾아 관리하도록 해야한다고 생각했습니다.
이때 무조건 기존의 객체 만을 활용해야 한다고 생각해버린 것이 문제였습니다.

그래서 좀 더 작은 단위로, 
이 상수가 관리되는 메소드가 무엇인지, 
그리고 이 상수가 선언된 이유와 결국 분리하자 선택한 이유를 생각하니

상수를 관리할 책임 자체가 이전의 객체들로부터 분리되어야 했던 것 같습니다.

jinyoungchoi95 commented on src/test/java/attendance/AttendanceBookTest.java:
테스트클래스는 테스트하고자하는 클래스와 동일한 패키지에 배치해주세요.

wodnd0131 replied:
제 코드의 Manager라는 이름을 가진 클래스들을 공통적으로, 책임이 너무나 혼재된 객체로 만든 것 같습니다.

리펙토링 과정에서 계층 구조를 모두 없애면서 이전의 로직을 따라 하려는 것이 잘못된 객체를 만든 것 같습니다.
게다가,, switch/case의 좋은 경우를 반영해보고자 다형성을 잘못 적용한 게 컸던 것 같습니다.
서비스 단위에 적용했다면 사용했을 법한데, 한정된 인터페이스로 서로 다른 도메인 객체를 구현하려니 
같은 인터페이스를 상속 받았던 Manager들이 늘어날 수록 코드가 산으로 가버렸습니다.😂

결국 상속을 포기했으나, 여전히 Manager 객체들이 너무 꼬여있어서 하나 씩 내부 책임을 분리하며, 
리펙토링하게 되었습니다. 테스트 단위와 객체의 책임에 대해 고민하고, 신경 쓸 수 있게 된 배경이 된 것 같습니다.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 몽이 😄

몇가지 코멘트 남겨두었으니 확인부탁드려요.

그리고 이후부터는 코멘트의 resolve 처리는 하지 않으셔도 됩니다. 반영 여부를 리뷰어가 확인하기 힘들뿐더러 pr 코멘트들은 다른 사람 혹은 나중에 몽이도 보게되는 학습자료가 되므로 미션내에서는 resolve 처리는 하지 않는 것이 좋습니다.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

Commit:
refactor: if문 가독성 관련 리펙터링

Commit:
refactor: beforeEach 초기화 제거

Commit:
refactor: 네이밍 개선

Commit:
refactor: stream 사용 개선

Commit:
refactor: getAttendanceList의 null 조건 제거

Commit:
refactor: common 패키지 제거

Commit:
refactor: manager 객체들의 manange() 메서드를 제거한다

Commit:
refactor: RegisterManagerTest를 AttendancesTest로 수정

-refactor: 테스트 코드를 세분화해 기능을 명확히 한다
-refactor: test를 위해 존재했던 findAttendance(name,attendance)를 삭제한다

Commit:
refactor: AttendancesTest로 수정하기 테스트를 이동

- 불필요해진 ModifyManagerTest를 삭제한다
- Attendance의 기능에 대한 추가적인 테스트는 불필요하므로, 삭제한다

Commit:
refactor: StatusStatistic으로부터 AttendancesHistory 분리

- StatusStatistic을 HistoryStatistic으로 이름을 수정한다
- Attendances를 record 클래스로, 그리고 list 대신 map을 사용하도록 수정한다

Commit:
refactor: AttendanceFileReader를 CsvReader로 변경

Commit:
refactor: StatusStatistic을 HistoryStatistic로 수정

- 관련된 테스트 또한 수정한다

Commit:
refactor: 상수 정리

Commit:
refactor: AttendanceStatus가 status를 결정하도록 수정

Commit:
refactor: 로직을 가진 enum 클래스 분리

-refactor: 불필요한 domain/attendance 패키지 제거

Commit:
refactor: 공통 변수를 모은 SystemDateConfig를 제거

-대신 SystemDateTime 객체를 주입받도록 수정

Commit:
refactor: int 파싱 예외 처리 추가

Commit:
refactor: OutputView에서 출력 형태의 책임을 가지도록 수정

Commit:
refactor: OutputView에서 파라미터로 받는 도메인 객체 제거

Commit:
refactor: InputValidator 리펙터링

Commit:
refactor: AttendanceBookFactory 추가

- AttendanceBook 생성 시 다른 도메인 객체를 직접 생성하지 않도록 팩토리 객체로 분리한다

Commit:
refactor: InputValidationException class 분리

Commit:
refactor: InputValidator 제거

Commit:
refactor: enum으로부터 출력값 제거

Commit:
refactor: EnumTextConverter 추가

-refactor:  enum이 displayable를 구현하도록 수정

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,29 @@
+public record AttendanceBook(Map<String, Attendances> attendances) {
+    private static final String NOT_REGISTERED_NICKNAME = "등록되지 않은 닉네임입니다.";
+
+    public Optional<Attendance> findAttendance(String nickname, LocalDate date) {
```
`Optional`을 반환하여 사용하도록 해야만하는지에 대해서 고민해보세요. 결과가 있고, 없고에 대한 처리를 외부에 넘기는데 책임이 분산되는 것 아닐까요?

wodnd0131 replied:
Optional에 대한 지식이 부족했던 것 같습니다.
Optional 대신, 무단 결석이라는 상태로 관리 가짜 객체를 추가하였습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceHistory.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.Map;
+import java.util.Optional;
+
+public record AttendanceHistory(Map<LocalDate, Optional<Attendance>> history) {
```
클래스 변수으로 Optional을 가지지 않습니다. 왜 Optional을 쓰게 되는지, 어떤 상황에서만 Optional을 써야하는지 고민해보세요.

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public static AttendanceStatus decideStatus(LocalTime time, LocalTime baseSchedule) {
+        for (AttendanceStatus status : values()) {
+            if (status.isAfterThreshold(time, baseSchedule)) {
+                return status;
+            }
+        }
```
> indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.

jinyoungchoi95 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,38 @@
+import attendance.exception.AttendanceArgumentException;
+
+public record Attendances(Map<LocalDate, Attendance> attendances) {
+    public static final String DUPLICATE_DATE = "이미 출석되었습니다. 수정 기능을 이용해주세요.";
```
```suggestion
    private static final String DUPLICATE_DATE = "이미 출석되었습니다. 수정 기능을 이용해주세요.";
```

jinyoungchoi95 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,38 @@
+        attendances.remove(date);
+    }
+
+    public void validateDuplicate(LocalDate date) {
+        if (attendances.containsKey(date)) {
+            throw new AttendanceArgumentException(DUPLICATE_DATE);
+        }
+    }
```
```java
        attendances.validateDuplicate(dateTime.toLocalDate());
        attendances.add(newAttendance);
```
외부에서 단일로 각각 호출하면 `Attendances`의 책임은 무엇인가요? 외부에서 순서대로 호출하지않으면 도메인이 잘 보호가 될까요?

wodnd0131 replied:
"객체에게 일을 시킨다" 라는 개념을 적용하는데 고민이 부족했네요.
단순히 로직을 객체 내부로 옮긴다는 것만 생각하여, 
특정한 행동을 시킨다 보단, 단순히 코드를 객체 안에 숨긴다 정도의 작업을 진행했던 것 같습니다.

리펙터링하면서 Attendances의 역할이 점점 모호해졌으나, 원인을 제대로 파악하지 못했습니다.
이러한 책임에 대한 고민이 부족했기 때문인 것 같습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/Displayable.java:
```diff
@@ -0,0 +1,5 @@
+package attendance.domain;
+
+public interface Displayable {
```
자바 클래스명은 명사가 컨벤션입니다.

https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

jinyoungchoi95 commented on src/main/java/attendance/domain/SanctionLevel.java:
```diff
@@ -0,0 +1,36 @@
+        return condition.test(wight);
+    }
+
+    public static SanctionLevel getValueByWight(int wight) {
```
음수가 입력가능한건 이상하지않을까요?

wodnd0131 replied:
로직에서 객체로 분리하면서, 기존의 로직에선 고려하지 않았던 부분도 마땅히 관리해줘야 하는 상황이었군요..
예외에 대한 생각이 부족했던 것 같습니다.

jinyoungchoi95 reviewed (approved):
안녕하세요 몽이 😄

다음 미션 진행을 위해 머지하겠습니다.
코멘트남겨둔 부분은 다음 단계에서 새로 구현할때 같이 고민하면서 반영해주세요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️
